File: packages/module/android/src/main/java/expo/modules/easypasskey/ExpoEasyPasskeyModule.kt
package expo.modules.easypasskey

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoEasyPasskeyModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoEasyPasskey")

    Function("isSupported") {
      PasskeyCeremonyAdapter.isSupported
    }

    Function("getPlatform") {
      "android"
    }

    AsyncFunction("create") Coroutine { options: Map<String, Any?> ->
      val activity = appContext.currentActivity ?: throw PasskeyMissingActivityException()
      PasskeyCeremonyAdapter(activity).create(PasskeyCreateRequest(options))
    }

    AsyncFunction("get") Coroutine { options: Map<String, Any?> ->
      val activity = appContext.currentActivity ?: throw PasskeyMissingActivityException()
      PasskeyCeremonyAdapter(activity).get(PasskeyGetRequest(options))
    }

    AsyncFunction("normalizeChallenge") { challenge: String ->
      PasskeyEncoding.normalizeBase64Url(challenge)
    }

    AsyncFunction("validateRelyingPartyId") { rpId: String ->
      PasskeyInputValidator.validateRpId(rpId)
    }

    AsyncFunction("describeCeremony") { kind: String, challenge: String, rpId: String, origin: String ->
      PasskeyRustHelpers.describeCeremony(kind, challenge, rpId, origin)
    }
  }
}

File: packages/module/ios/ExpoEasyPasskeyModule.swift
import AuthenticationServices
import ExpoModulesCore

public class ExpoEasyPasskeyModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoEasyPasskey")

    Function("isSupported") {
      ExpoEasyPasskeyCeremonyAdapter.isSupported
    }

    Function("getPlatform") {
      "ios"
    }

    AsyncFunction("create") { (_ options: [String: Any]) async throws -> [String: Any] in
      let request = try PasskeyCreateRequest(options)
      let policy = try PasskeyRegistrationPolicy.resolve(request)
      let adapter = try await self.makeCeremonyAdapter()
      return try await adapter.create(request, policy: policy)
    }

    AsyncFunction("get") { (_ options: [String: Any]) async throws -> [String: Any] in
      let request = try PasskeyGetRequest(options)
      let adapter = try await self.makeCeremonyAdapter()
      return try await adapter.get(request)
    }

    AsyncFunction("normalizeChallenge") { (challenge: String) async throws -> String in
      try PasskeyEncoding.normalizeBase64Url(challenge)
    }

    AsyncFunction("validateRelyingPartyId") { (rpId: String) async throws -> String in
      try PasskeyInputValidator.validateRpId(rpId)
    }

    AsyncFunction("describeCeremony") { (
      kind: String,
      challenge: String,
      rpId: String,
      origin: String
    ) async throws -> [String: Any] in
      try PasskeyRustHelpers.describeCeremony(
        kind: kind,
        challenge: challenge,
        rpId: rpId,
        origin: origin
      )
    }
  }

  @MainActor
  private func makeCeremonyAdapter() throws -> ExpoEasyPasskeyCeremonyAdapter {
    guard let viewController = appContext?.utilities?.currentViewController() else {
      throw PasskeyMissingPresentationContextException()
    }

    return ExpoEasyPasskeyCeremonyAdapter(presentationAnchorProvider: {
      viewController.view.window ?? UIWindow()
    })
  }
}
